### PR TITLE
The the permutedims compiling problem for high dimensional arrays

### LINKED ISSUE #42438: Unreasonable slow just in time compiling of `permutedims` when the array dimension is high
Permuting the dimensions of high dimensional tensors is frequently used in tensor networks.

```julia
julia> using Random, Profile

julia> (n=24; t = randn(fill(2, n)...); @time permutedims(t, randperm(n)));
 39.144106 seconds (1.21 M allocations: 184.261 MiB, 0.09% gc time, 99.45% compilation time)

julia> (n=24; t = randn(fill(2, n)...); @time permutedims(t, randperm(n)));
  0.194273 seconds (9 allocations: 128.001 MiB, 25.46% gc time)

julia> (n=28; t = randn(fill(2, n)...); @time permutedims(t, randperm(n)));
657.982409 seconds (1.04 M allocations: 2.044 GiB, 0.00% gc time, 99.20% compilation time)

julia> (n=28; t = randn(fill(2, n)...); @time permutedims(t, randperm(n)));
  6.854463 seconds (3.84 k allocations: 2.000 GiB, 1.76% gc time, 0.07% compilation time)
```

This time is much longer than `TensorOperations.tensorcopy` (~6s).
The crazy fact (from my observation) is: the compiling time of `permutedims` scales exponentially with the tensor rank!

Related code block: https://github.com/JuliaLang/julia/blob/93d375cb021379a7a57aad05df86c5925cfaebcf/base/multidimensional.jl#L1532
Julia version: 1.7.0-beta4

The problem might be related to the compiling of multi-loops.

@kshyatt

### PR DESCRIPTION
**Pull request by GiggleLiu**
This PR fixes #42438. I do not know why it is more friendly to the compiler, I just some try and errors. Someone might want to look deeper into this issue (might be in LLVM according to @timholy 's comment), but we need to fix this issue due to the the need in some practical applications.

```julia
using Base.Cartesian
using Base: size_to_strides, checkdims_perm
using Random

for (V, PT, BT) in Any[((:N,), BitArray, BitArray), ((:T,:N), Array, StridedArray)]
    @eval @generated function newperm!(P::$PT{$(V...)}, B::$BT{$(V...)}, perm) where $(V...)
        quote
            checkdims_perm(P, B, perm)

            #calculates all the strides
            native_strides = size_to_strides(1, size(B)...)
            strides_1 = 0
            @nexprs $N d->(strides_{d+1} = native_strides[perm[d]])

            #Creates offset, because indexing starts at 1
            offset = 1 - sum(@ntuple $N d->strides_{d+1})

            sumc = 0
            ind = 1
            @nexprs 1 d->(counts_{$N+1} = strides_{$N+1}) # a trick to set counts_($N+1)
            @nloops($N, i, P,
                    d->(df_d=i_d*strides_{d+1} ;sumc += df_d), # PRE
                    d->(sumc -= df_d), # POST
                    begin # BODY
                        @inbounds P[ind] = B[sumc+offset]
                        ind += 1
                    end)

            return P
        end
    end
end


using Test

@testset "newperm" begin
    n=25
    t=randn(rand(1:2, n)...)
    perm = randperm(n)
    p = zeros(eltype(t), size.(Ref(t), (perm...,)));
    @time newperm!(p, t, perm);
    # 0.395072 seconds (520.17 k allocations: 20.894 MiB, 99.99% compilation time)
    @time permutedims!(p, t, perm);
    # 41.520901 seconds (502.11 k allocations: 20.155 MiB, 100.00% compilation time)
    @test newperm!(p, t, perm) ≈ permutedims!(p, t, perm)
end

using BenchmarkTools
# high dim
n=25
t=randn(rand(2:2, n)...)
perm = randperm(n)
p = zeros(eltype(t), size.(Ref(t), (perm...,)))
@btime newperm!($p, $t, $perm)
# 179.814 ms (2 allocations: 96 bytes)
@btime permutedims!($p, $t, $perm)
# 201.038 ms (2 allocations: 96 bytes)

# low dim (make sure no performance regression)
t=randn(100, 200, 300)
perm = [3,2,1]
p = zeros(eltype(t), size.(Ref(t), (perm...,)))
@btime newperm!($p, $t, $perm)
# 19.203 ms (2 allocations: 96 bytes)
@btime permutedims!($p, $t, $perm);
# 19.339 ms (2 allocations: 96 bytes)
```

### CODE REVIEW

**Comment by GiggleLiu** [comment]
(sorry for some unrelated commit history, whoever reviews this PR should use squash and merge)

File: base/multidimensional.jl [hunk 1]
```diff
@@ -1577,13 +1577,13 @@ for (V, PT, BT) in Any[((:N,), BitArray, BitArray), ((:T,:N), Array, StridedArra
             #Creates offset, because indexing starts at 1
             offset = 1 - sum(@ntuple $N d->strides_{d+1})
 
+            sumc = 0
             ind = 1
             @nexprs 1 d->(counts_{$N+1} = strides_{$N+1}) # a trick to set counts_($N+1)
```
Comment (vtjnash) [inline base/multidimensional.jl:1581]:
```suggestion
```

**Review by vtjnash — COMMENTED** [review]
SGTM

File: base/multidimensional.jl (hunk 1 above)
Comment (GiggleLiu) [reply base/multidimensional.jl:1581]: 👍🏾 , removed.

**Review by timholy — APPROVED** [review]

**Pull request merged**